{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Beginning installation. This may take up to 40 minutes to complete.\n",
      "The transcriptional reference data for GRCh38 has been saved.\n",
      "All reference files have been created.\n",
      "Verifying and benchmarking installation now...\n",
      "Starting matrix generation for SNVs and DINUCs...Completed! Elapsed time: 1.83 seconds.\n",
      "Matrices generated for 1 samples with 0 errors. Total of 9631 SNVs, 0 DINUCs, and 0 INDELs were successfully analyzed.\n",
      "Installation was succesful.\n",
      "SigProfilerMatrixGenerator took 1.884936809539795 seconds to complete.\n",
      "To proceed with matrix_generation, please provide the path to your vcf files and an appropriate output path.\n",
      "Installation complete.\n"
     ]
    }
   ],
   "source": [
    "from SigProfilerMatrixGenerator import install as genInstall\n",
    "genInstall.install('GRCh38', rsync=False, bash=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting matrix generation for SNVs and DINUCs...Completed! Elapsed time: 268.01 seconds.\n",
      "Starting matrix generation for INDELs...Completed! Elapsed time: 108.11 seconds.\n",
      "Matrices generated for 28 samples with 56101 errors. Total of 324506 SNVs, 15738 DINUCs, and 236863 INDELs were successfully analyzed.\n"
     ]
    }
   ],
   "source": [
    "from SigProfilerMatrixGenerator.scripts import SigProfilerMatrixGeneratorFunc as matGen\n",
    "\n",
    "matrices = matGen.SigProfilerMatrixGeneratorFunc(\"Scitron\", \n",
    "                                                 \"GRCh38\", \n",
    "                                                 \"/Users/jfnavarro/Projects/Scitron/Results/Mutect2VCF/\",\n",
    "                                                 plot=True, \n",
    "                                                 exome=True, \n",
    "                                                 bed_file='/Users/jfnavarro/Projects/Scitron/intervals/intersect_collapsed.bed', \n",
    "                                                 chrom_based=False, \n",
    "                                                 tsb_stat=False, \n",
    "                                                 seqInfo=False, \n",
    "                                                 cushion=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['6144', '384', '1536', '96', '6', '24', '4608', '288', '18', 'DINUC', 'ID'])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrices.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************** Reported Current Memory Use: 9.88 GB *****************\n",
      "\n",
      "Starting matrix generation for SNVs and DINUCs...Completed! Elapsed time: 15.19 seconds.\n",
      "Starting matrix generation for INDELs...Completed! Elapsed time: 36.61 seconds.\n",
      "Matrices generated for 28 samples with 56101 errors. Total of 324506 SNVs, 15738 DINUCs, and 236863 INDELs were successfully analyzed.\n",
      "Extracting signature 1 for mutation type 96\n",
      "The matrix normalizig cutoff is 12475\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 1 signatures is 242.74 seconds\n",
      "Optimization time is 3.9178121089935303 seconds\n",
      "The reconstruction error is 0.3792, average process stability is 1.0 and \n",
      "the minimum process stability is 1.0 for 1 signatures\n",
      "\n",
      "\n",
      "Extracting signature 2 for mutation type 96\n",
      "The matrix normalizig cutoff is 12475\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 2 signatures is 281.29 seconds\n",
      "Optimization time is 3.3794760704040527 seconds\n",
      "The reconstruction error is 0.2431, average process stability is 1.0 and \n",
      "the minimum process stability is 1.0 for 2 signatures\n",
      "\n",
      "\n",
      "Extracting signature 3 for mutation type 96\n",
      "The matrix normalizig cutoff is 12475\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 3 signatures is 327.5 seconds\n",
      "Optimization time is 3.764653205871582 seconds\n",
      "The reconstruction error is 0.226, average process stability is 0.92 and \n",
      "the minimum process stability is 0.89 for 3 signatures\n",
      "\n",
      "\n",
      "Extracting signature 4 for mutation type 96\n",
      "The matrix normalizig cutoff is 12475\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 4 signatures is 375.29 seconds\n",
      "Optimization time is 3.1468520164489746 seconds\n",
      "The reconstruction error is 0.1743, average process stability is 0.96 and \n",
      "the minimum process stability is 0.94 for 4 signatures\n",
      "\n",
      "\n",
      "Extracting signature 5 for mutation type 96\n",
      "The matrix normalizig cutoff is 12475\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 5 signatures is 608.42 seconds\n",
      "Optimization time is 3.1158971786499023 seconds\n",
      "The reconstruction error is 0.1203, average process stability is 0.94 and \n",
      "the minimum process stability is 0.89 for 5 signatures\n",
      "\n",
      "\n",
      "Decompositon Plot made for SBS96A\n",
      "\n",
      "Decompositon Plot made for SBS96B\n",
      "\n",
      "Decompositon Plot made for SBS96C\n",
      "\n",
      "Decompositon Plot made for SBS96D\n",
      "\n",
      "Extracting signature 1 for mutation type DINUC\n",
      "The matrix normalizig cutoff is 7800\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 1 signatures is 226.79 seconds\n",
      "Optimization time is 3.197880983352661 seconds\n",
      "The reconstruction error is 0.5052, average process stability is 1.0 and \n",
      "the minimum process stability is 1.0 for 1 signatures\n",
      "\n",
      "\n",
      "Extracting signature 2 for mutation type DINUC\n",
      "The matrix normalizig cutoff is 7800\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 2 signatures is 266.13 seconds\n",
      "Optimization time is 3.1636319160461426 seconds\n",
      "The reconstruction error is 0.246, average process stability is 0.99 and \n",
      "the minimum process stability is 0.99 for 2 signatures\n",
      "\n",
      "\n",
      "Extracting signature 3 for mutation type DINUC\n",
      "The matrix normalizig cutoff is 7800\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 3 signatures is 366.94 seconds\n",
      "Optimization time is 3.132366895675659 seconds\n",
      "The reconstruction error is 0.1944, average process stability is 0.95 and \n",
      "the minimum process stability is 0.89 for 3 signatures\n",
      "\n",
      "\n",
      "Extracting signature 4 for mutation type DINUC\n",
      "The matrix normalizig cutoff is 7800\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 4 signatures is 494.18 seconds\n",
      "Optimization time is 3.2188849449157715 seconds\n",
      "The reconstruction error is 0.1751, average process stability is 0.79 and \n",
      "the minimum process stability is 0.5 for 4 signatures\n",
      "\n",
      "\n",
      "Extracting signature 5 for mutation type DINUC\n",
      "The matrix normalizig cutoff is 7800\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 5 signatures is 630.01 seconds\n",
      "Optimization time is 3.13753080368042 seconds\n",
      "The reconstruction error is 0.151, average process stability is 0.56 and \n",
      "the minimum process stability is -0.01 for 5 signatures\n",
      "\n",
      "\n",
      "Decompositon Plot made for DBS78A\n",
      "\n",
      "Decompositon Plot made for DBS78B\n",
      "\n",
      "Decompositon Plot made for DBS78C\n",
      "\n",
      "Extracting signature 1 for mutation type INDEL\n",
      "The matrix normalizig cutoff is 8809\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 1 signatures is 228.47 seconds\n",
      "Optimization time is 3.165351152420044 seconds\n",
      "The reconstruction error is 0.739, average process stability is 1.0 and \n",
      "the minimum process stability is 1.0 for 1 signatures\n",
      "\n",
      "\n",
      "Extracting signature 2 for mutation type INDEL\n",
      "The matrix normalizig cutoff is 8809\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 2 signatures is 271.36 seconds\n",
      "Optimization time is 3.1527609825134277 seconds\n",
      "The reconstruction error is 0.6567, average process stability is 0.91 and \n",
      "the minimum process stability is 0.88 for 2 signatures\n",
      "\n",
      "\n",
      "Extracting signature 3 for mutation type INDEL\n",
      "The matrix normalizig cutoff is 8809\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 3 signatures is 282.75 seconds\n",
      "Optimization time is 3.1510989665985107 seconds\n",
      "The reconstruction error is 0.1722, average process stability is 1.0 and \n",
      "the minimum process stability is 0.99 for 3 signatures\n",
      "\n",
      "\n",
      "Extracting signature 4 for mutation type INDEL\n",
      "The matrix normalizig cutoff is 8809\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 4 signatures is 393.78 seconds\n",
      "Optimization time is 3.161674976348877 seconds\n",
      "The reconstruction error is 0.1015, average process stability is 0.99 and \n",
      "the minimum process stability is 0.97 for 4 signatures\n",
      "\n",
      "\n",
      "Extracting signature 5 for mutation type INDEL\n",
      "The matrix normalizig cutoff is 8809\n",
      "\n",
      "\n",
      "Time taken to collect 500 iterations for 5 signatures is 727.73 seconds\n",
      "Optimization time is 3.1800830364227295 seconds\n",
      "The reconstruction error is 0.0905, average process stability is 0.94 and \n",
      "the minimum process stability is 0.74 for 5 signatures\n",
      "\n",
      "\n",
      "Decompositon Plot made for ID83A\n",
      "\n",
      "Decompositon Plot made for ID83B\n",
      "\n",
      "Decompositon Plot made for ID83C\n",
      "\n",
      "Decompositon Plot made for ID83D\n",
      "\n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfiler Extractor.\n",
      " \n"
     ]
    }
   ],
   "source": [
    "from SigProfilerExtractor import sigpro as sig\n",
    "\n",
    "data = \"/Users/jfnavarro/Projects/Scitron/Results/Mutect2VCF/\"\n",
    "sig.sigProfilerExtractor(\"vcf\", \"sig_profiler-output\", data, \n",
    "                         exome=True, reference_genome=\"GRCh38\", \n",
    "                         minimum_signatures=1, maximum_signatures=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
