{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "mutations = pd.read_csv('../analysis/merged_variants_anno.txt', sep=\"\\t\", header=0, index_col=0)\n",
    "metasamples = pd.read_excel(\"../Metadata.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert matrix to unique VAF value per cell (mean of VAFs)\n",
    "import numpy as np\n",
    "\n",
    "FILTER_DP = 25\n",
    "FILTER_VAF = 5.0\n",
    "FILTER_EFFECTS = ['chromosome', \n",
    "                  'duplication',\n",
    "                  'inversion',\n",
    "                  'exon_loss_variant',\n",
    "                  'frameshift_variant',\n",
    "                  'missense_variant',\n",
    "                  'protein_protein_contact',\n",
    "                  'structural_interaction_variant',\n",
    "                  'splice_acceptor_variant',\n",
    "                  'splice_donor_variant',\n",
    "                  'stop_lost', \n",
    "                  'start_lost', \n",
    "                  'stop_gained']\n",
    "\n",
    "def unify_vaf(str_vaf):\n",
    "    if str_vaf in ['Na', 'NA']:\n",
    "        return 0.0\n",
    "    values = [float(c.split(':')[1].split(';')[2]) \n",
    "              if int(c.split(':')[1].split(';')[0]) >= FILTER_DP else 0.0 for c in str_vaf.split('|')]\n",
    "    merged_vaf = np.round(np.mean(values), 2)\n",
    "    return merged_vaf if merged_vaf >= FILTER_VAF else 0.0\n",
    "\n",
    "def unify_vaf_unfilter(str_vaf):\n",
    "    if str_vaf in ['Na', 'NA']:\n",
    "        return 0.0\n",
    "    values = [float(c.split(':')[1].split(';')[2]) for c in str_vaf.split('|')]\n",
    "    return np.round(np.mean(values), 2)\n",
    "\n",
    "def get_maf(ref, alt):\n",
    "    ref_len = len(ref)\n",
    "    alt_len = len(alt)\n",
    "    diff = ref_len - alt_len\n",
    "    vtype = None\n",
    "    if diff == 0:\n",
    "        if alt_len == 1:\n",
    "            vtype = 'SNP'\n",
    "        elif alt_len == 2:\n",
    "            vtype = 'DNP'\n",
    "        elif alt_len == 3:\n",
    "            vtype = 'TNP'\n",
    "        else:\n",
    "            vtype = 'ONP'\n",
    "    elif diff > 0 and alt_len == 1:\n",
    "        vtype = 'DEL'\n",
    "    elif diff < 0 and ref_len == 1:\n",
    "        vtype = 'INS'\n",
    "    else:\n",
    "        vtype = 'INDEL'\n",
    "    return vtype\n",
    "\n",
    "\n",
    "# Convert vcalls to a single VAF (no DP filter)\n",
    "mutations_vaf = mutations.copy()\n",
    "mutations_vaf.loc[:, mutations_vaf.columns[14:]] = mutations_vaf.loc[:, mutations_vaf.columns[14:]].applymap(unify_vaf_unfilter)\n",
    "mutations_vaf.to_csv('../analysis/merged_variants_anno_vaf.txt', sep=\"\\t\", index=True, header=True)\n",
    "\n",
    "# Convert vcalls to a single VAF (with DP filter)\n",
    "mutations_vaf = mutations.copy()\n",
    "mutations_vaf.loc[:, mutations_vaf.columns[14:]] = mutations_vaf.loc[:, mutations_vaf.columns[14:]].applymap(unify_vaf)\n",
    "\n",
    "\n",
    "# Create a MAF matrix (applying filters with GNOMAD filter)\n",
    "to_keep = list()\n",
    "maf_df = pd.DataFrame(columns=['Hugo_Symbol',\n",
    "                               'Tumor_Sample_Barcode',\n",
    "                               'Chromosome',\n",
    "                               'Start_Position',\n",
    "                               'Reference_Allele',\n",
    "                               'Tumor_Seq_Allele2',\n",
    "                               'Variant_Type'])\n",
    "for index, row in mutations_vaf.iterrows():\n",
    "    has_cosmic = row['COSMI70'] not in ['Na', '.']\n",
    "    has_dbsnp = row['DBSNP'] not in ['Na', '.']\n",
    "    has_gnomad = row['GNOMAD'] not in ['Na', '.']\n",
    "    has_clinvar = 'Pathogenic' in row['CLINVAR']\n",
    "    gnomad = float(row['GNOMAD']) if has_gnomad else 1.0\n",
    "    \n",
    "    if has_gnomad and gnomad >= 0.01:\n",
    "        continue\n",
    "    \n",
    "    effect = row['EFFECT']\n",
    "    vkey = index\n",
    "    gene = row['GENE']\n",
    "    chrm, rest = vkey.split(':')\n",
    "    start, rest2 = rest.split()\n",
    "    ref, alt = rest2.split('>')\n",
    "    start = int(start)\n",
    "    if any(x in FILTER_EFFECTS for y in effect.split('|') for x in y.split('&')):\n",
    "        vtype = get_maf(ref, alt)\n",
    "        if vtype == 'DEL':\n",
    "            alt = '-' if len(alt) == 1 else alt[1:]\n",
    "            ref = ref[1:]\n",
    "            start += 1\n",
    "        elif vtype == 'INS':\n",
    "            ref = '-' if len(ref) == 1 else ref[1:]\n",
    "            alt = alt[1:]\n",
    "            start += 1\n",
    "        has_mutations = False\n",
    "        for column, value in row[14:].items():\n",
    "            patient = column.split('-')[-1]\n",
    "            tumor_only = not any(metasamples[metasamples['PATIENT_ID'] == patient]['SAMPLE_TYPE'] == 'gDNA')\n",
    "            if tumor_only and not has_clinvar and not has_cosmic:\n",
    "                continue\n",
    "    \n",
    "            if float(value) >= FILTER_VAF and 'gDNA' not in column:\n",
    "                maf_df = maf_df.append({'Hugo_Symbol':gene,\n",
    "                                        'Tumor_Sample_Barcode':column,\n",
    "                                        'Chromosome':chrm,\n",
    "                                        'Start_Position':start,\n",
    "                                        'Reference_Allele':ref,\n",
    "                                        'Tumor_Seq_Allele2':alt,\n",
    "                                        'Variant_Type':vtype},\n",
    "                                       ignore_index=True)\n",
    "                has_mutations = True\n",
    "        if has_mutations:\n",
    "            to_keep.append(index)\n",
    "                \n",
    "mutations_vaf = mutations_vaf.reindex(to_keep)\n",
    "mutations_vaf = mutations_vaf.drop(['NUM_SAMPLES', \n",
    "                                    'NUM_SAMPLES_GENE',\n",
    "                                    'NUM_PATIENTS',\n",
    "                                    'NUM_PATIENTS_GENE'], \n",
    "                                    axis='columns')\n",
    "mutations_vaf.to_csv('../analysis/merged_variants_anno_vaf_filtered.txt', sep=\"\\t\", index=True, header=True)\n",
    "\n",
    "maf_df['Start_Position'] = pd.to_numeric(maf_df['Start_Position'])\n",
    "maf_df.to_csv('../analysis/merged_maf.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import signatureanalyzer as sa\n",
    "\n",
    "maf_df = pd.read_csv('../analysis/merged_maf.txt', sep=\"\\t\", index_col=False, header=0)\n",
    "\n",
    "_,spectra_snvs = sa.spectra.get_spectra_from_maf(maf_df, cosmic='cosmic3_exome', hgfile='../analysis/hg38.2bit')\n",
    "spectra_snvs.to_csv('../analysis/merged_maf_spectra_snvs.txt', sep=\"\\t\", index=True, header=True)\n",
    "_,spectra_indels = sa.spectra.get_spectra_from_maf(maf_df, cosmic='cosmic3_ID', hgfile='../analysis/hg38.2bit')\n",
    "spectra_indels.to_csv('../analysis/merged_maf_spectra_indels.txt', sep=\"\\t\", index=True, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
