{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 12 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 140 samples\n",
      "Searching for CNVs in ../Results/VariantCalling\n",
      "../Results/VariantCalling/1123_3_cfDNA_vs_B19_0971__17_22863_A1__0_gDNA/Control-FREEC/1276_23_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/1261_2_cfDNA_vs_B19_0971__17_22863_A1__0_gDNA/Control-FREEC/1261_2_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0970__VH_17_B_022863_A5__1_tumor_vs_B19_0971__17_22863_A1__0_gDNA/Control-FREEC/B19_0970__VH_17_B_022863_A5__1_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/1514_5_cfDNA_vs_B190982_36_gDNA/Control-FREEC/1514_5_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/916_6_cfDNA_vs_B190982_36_gDNA/Control-FREEC/916_6_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0972__16_39495_A1__4_tumor_vs_B190982_36_gDNA/Control-FREEC/B19_0972__16_39495_A1__4_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/1567_8_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/1975_7_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/2418_10_cfDNA_vs_L19_5550_gDNA/Control-FREEC/2418_10_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/3031_11_cfDNA_vs_L19_5550_gDNA/Control-FREEC/3031_11_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX090_PDX_vs_L19_5550_gDNA/Control-FREEC/CTAX090_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/186_132_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/216_14_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/2403_17_cfDNA/Control-FREEC/2403_17_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/2985_18_cfDNA/Control-FREEC/2985_18_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0973__M18_2588__16_tumor_vs_B1970981_15_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0973__M18_2588__16_tumor/Control-FREEC/B19_0973__M18_2588__16_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/1276_23_cfDNA/Control-FREEC/1276_23_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/2205_22_cfDNA/Control-FREEC/2205_22_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0974__M17_2775__21_tumor/Control-FREEC/B19_0974__M17_2775__21_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_975__18_33915__20_tumor/Control-FREEC/B19_975__18_33915__20_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/286_27_cfDNA/Control-FREEC/286_27_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/327_28_cfDNA/Control-FREEC/327_28_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0976__13_105040__26_tumor/Control-FREEC/B19_0976__13_105040__26_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/1298_30_cfDNA/Control-FREEC/1298_30_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/934_31_cfDNA/Control-FREEC/934_31_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/M16-5225-EBP/Control-FREEC/M16-5225-EBP.pileup_ratio.txt\n",
      "../Results/VariantCalling/1033_35_cfDNA_vs_B190986_RG1297_32_gDNA/Control-FREEC/1033_35_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/1297_34_cfDNA/Control-FREEC/1297_34_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/1297_34_cfDNA_vs_B190986_RG1297_32_gDNA/Control-FREEC/1297_34_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0979__M17_1352__33_tumor/Control-FREEC/B19_0979__M17_1352__33_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0979__M17_1352__33_tumor_vs_B190986_RG1297_32_gDNA/Control-FREEC/B19_0979__M17_1352__33_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/2237_39_cfDNA/Control-FREEC/2237_39_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/2993_38_cfDNA/Control-FREEC/2993_38_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/B19_0980__VH_18_B_008565_A13__37_tumor/Control-FREEC/B19_0980__VH_18_B_008565_A13__37_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX096_PDX/Control-FREEC/CTAX096_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/19B0050298_A27__B2010104__41_tumor_vs_3110_40_gDNA/Control-FREEC/19B0050298_A27__B2010104__41_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/3110_42_cfDNA_vs_3110_40_gDNA/Control-FREEC/3110_42_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/18_39888_A4__B2010101__43_tumor_vs_18-39888A8-MPGA/Control-FREEC/18_39888_A4__B2010101__43_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/2550_44_cfDNA_vs_18-39888A8-MPGA/Control-FREEC/2550_44_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/2762_47_cfDNA_vs_2762_45_gDNA/Control-FREEC/2762_47_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/M19_1039__B2010102__46_tumor_vs_2762_45_gDNA/Control-FREEC/M19_1039__B2010102__46_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/18_1801_A8__B2010100__48_tumor_vs_MGM_gDNA/Control-FREEC/18_1801_A8__B2010100__48_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/3131_50_cfDNA_vs_RMM_gDNA/Control-FREEC/3131_50_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/M19_593__B2010103__49_tumor_vs_RMM_gDNA/Control-FREEC/M19_593__B2010103__49_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/1132_52_cfDNA_vs_1260_AIA_gDNA/Control-FREEC/1132_52_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/M17_2060__B2010105__51_tumor_vs_1260_AIA_gDNA/Control-FREEC/M17_2060__B2010105__51_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/1455_54_cfDNA_vs_1292_RPC_Gdna/Control-FREEC/1455_54_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/977_55_cfDNA_vs_1292_RPC_Gdna/Control-FREEC/977_55_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/M17_615__B2010107__53_tumor_vs_1292_RPC_Gdna/Control-FREEC/M17_615__B2010107__53_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/1000_57_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/M17_723__B2010106__56_tumor/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/1632_59_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/M18_100__B2010110__58_tumor/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/VH_18_B_034996_60_tumor_vs_AC_gDNA/Control-FREEC/VH_18_B_034996_60_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/M19_2477_61_tumor_vs_4067EM/Control-FREEC/M19_2477_61_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/20_13032_A1_63_tumor_vs_4066_FRC_131__62_gDNA/Control-FREEC/20_13032_A1_63_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/4066_64_cfDNA_vs_4066_FRC_131__62_gDNA/Control-FREEC/4066_64_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX099_PDX_vs_4066_FRC_131__62_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/4113_67_cfDNA_vs_4113_65_gDNA/Control-FREEC/4113_67_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/VH___20___B___001520_66_tumor_vs_4113_65_gDNA/Control-FREEC/VH___20___B___001520_66_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/M19_2042_69_tumor/Control-FREEC/M19_2042_69_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/M20-1981-JVC_vs_JLCR_gDNA/Control-FREEC/M20-1981-JVC.pileup_ratio.txt\n",
      "../Results/VariantCalling/4150_L_vs_JLCR_gDNA/Control-FREEC/4150_L.pileup_ratio.txt\n",
      "../Results/VariantCalling/4169_73_cfDNA_vs_4169_FNM_132_71_gDNA/Control-FREEC/4169_73_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX087_PDX_vs_4169_FNM_132_71_gDNA/Control-FREEC/CTAX087_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX093_PDX_vs_4169_FNM_132_71_gDNA/Control-FREEC/CTAX093_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/4179_77_cfDNA_vs_4179_JSN_172_76_gDNA/Control-FREEC/4179_77_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/4213_79_cfDNA_vs_4213_CV_272__78_gDNA/Control-FREEC/4213_79_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/M20-920-JLCR_vs_4213_CV_272__78_gDNA/Control-FREEC/M20-920-JLCR.pileup_ratio.txt\n",
      "../Results/VariantCalling/M18_2959_80_tumor_vs_MDL_gDNA/Control-FREEC/M18_2959_80_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/JAM_1031_85_cfDNA_vs_JAM_1283_81_gDNA/Control-FREEC/JAM_1031_85_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/JAM_1207_86_cfDNA_vs_JAM_1283_81_gDNA/Control-FREEC/JAM_1207_86_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/JAM_1283_84_cfDNA_vs_JAM_1283_81_gDNA/Control-FREEC/JAM_1283_84_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/JAM_16_19733_A10_MPP216_L18_4331_82_tumor_vs_JAM_1283_81_gDNA/Control-FREEC/JAM_16_19733_A10_MPP216_L18_4331_82_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/JAM_776_83_cfDNA_vs_JAM_1283_81_gDNA/Control-FREEC/JAM_776_83_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/JAM_826_88_cfDNA_vs_JAM_1283_81_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/JAM_941_87_cfDNA_vs_JAM_1283_81_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../Results/VariantCalling/CTAX033_PDX_vs_L196746_gDNA/Control-FREEC/CTAX033_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_1022_95_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_1022_95_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_1065_90_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_1065_90_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_1167_94_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_1167_94_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_16_658947_MPP216_L18_4333_89_tumor_vs_L196746_gDNA/Control-FREEC/DAS_16_658947_MPP216_L18_4333_89_tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_529_93_cfDNA_vs_L196746_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_634_100_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_634_100_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_682_99_cfDNA_vs_L196746_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_721_91_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_721_91_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_795_98_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_795_98_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_843_97_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_843_97_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_861_96_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_861_96_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/DAS_901_92_cfDNA_vs_L196746_gDNA/Control-FREEC/DAS_901_92_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/IGO_1109_105_cfDNA_vs_IGO_1273_101_gDNA/Control-FREEC/IGO_1109_105_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/IGO_1273_103_cfDNA_vs_IGO_1273_101_gDNA/Control-FREEC/IGO_1273_103_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/IGO_687_108_cfDNA_vs_IGO_1273_101_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/IGO_784_107_cfDNA_vs_IGO_1273_101_gDNA/Control-FREEC/IGO_784_107_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/IGO_869_104_cfDNA_vs_IGO_1273_101_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/IGO_976_106_cfDNA_vs_IGO_1273_101_gDNA/Control-FREEC/tumor.pileup_ratio.txt\n",
      "../Results/VariantCalling/PDX_IGO_E200_109_PDX_vs_IGO_1273_101_gDNA/Control-FREEC/PDX_IGO_E200_109_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX041_PDX_vs_IGO_1273_101_gDNA/Control-FREEC/CTAX041_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/ARM_621_114_cfDNA_vs_L19_6477_gDNA/Control-FREEC/ARM_621_114_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/ARM_786_113_cfDNA_vs_L19_6477_gDNA/Control-FREEC/ARM_786_113_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/ARM_867_111_cfDNA_vs_L19_6477_gDNA/Control-FREEC/ARM_867_111_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/ARM_960_112_cfDNA_vs_L19_6477_gDNA/Control-FREEC/ARM_960_112_cfDNA.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX035_PDX_vs_L19_6477_gDNA/Control-FREEC/CTAX035_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/MCB_522_117_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/MCB_690_119_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/MCB_809_118_cfDNA/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/MCB_M16_1190_MPP216_L18_4332_116_tumor/Control-FREEC/sample.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX051_PDX_vs_L20_0748_gDNA/Control-FREEC/CTAX051_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX023_PDX_vs_L20_0747_gDNA/Control-FREEC/CTAX023_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX039_PDX_vs_L19_6479_gDNA/Control-FREEC/CTAX039_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/CTAX002_PDX_vs_L19_5546_gDNA/Control-FREEC/CTAX002_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/T074_PDX_vs_L19_5548_gDNA/Control-FREEC/T074_PDX.pileup_ratio.txt\n",
      "../Results/VariantCalling/T160_PDX_vs_L19_5549_gDNA/Control-FREEC/T160_PDX.pileup_ratio.txt\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "from pyensembl import EnsemblRelease\n",
    "\n",
    "# release 86 (GRCh38) was the one used in sarek\n",
    "data = EnsemblRelease(86)\n",
    "\n",
    "SAMPLES = \"../sarek/samples_all.tsv\"\n",
    "PATH = \"../Results/VariantCalling\"\n",
    "\n",
    "metasamples = pd.read_csv(SAMPLES, sep=\"\\t\", header=None)\n",
    "metasamples.columns = ['PATIENT_ID', 'GENDER', \"TUMOR\", \"SAMPLE_ID\", \"LANE\", \"FASTQ1\", \"FASTQ2\"]\n",
    "metasamples = metasamples.drop_duplicates(subset=['SAMPLE_ID'])\n",
    "\n",
    "print('Loaded {} samples'.format(metasamples.shape[0]))\n",
    "print('Searching for CNVs in {}'.format(PATH))\n",
    "\n",
    "# FIRST Parse CNVs from collapsed files\n",
    "cnvs_chr = defaultdict(lambda: defaultdict(float))\n",
    "for index, row in metasamples.iterrows():\n",
    "    sample_id = row['SAMPLE_ID']\n",
    "    sample_patient = row['PATIENT_ID']\n",
    "    tumor_only = not any(metasamples[metasamples['PATIENT_ID'] == sample_patient]['TUMOR'] == 0)\n",
    "    # Retrieve CNVs\n",
    "    if tumor_only:\n",
    "        for elem in Path(PATH).rglob('{}*/Control-FREEC/*pileup_ratio.txt'.format(sample_id)):\n",
    "            with open(elem, 'r') as filehandler:\n",
    "                print(elem)\n",
    "                for line in filehandler.readlines()[1:]:\n",
    "                    tokens = line.split('\\t')\n",
    "                    chrm = str(tokens[0])\n",
    "                    start = int(tokens[1])\n",
    "                    ratio = float(tokens[2])\n",
    "                    # Control-FREEC defines a -1 ratio as not present\n",
    "                    if ratio != -1:\n",
    "                        cv = int(tokens[4])\n",
    "                        cnvs_chr['{}:{}'.format(chrm, start)][sample_id] = (cv,ratio)\n",
    "    else:\n",
    "        gdna = metasamples[(metasamples['PATIENT_ID'] == sample_patient)\n",
    "                           & (metasamples['TUMOR'] == 0)]['SAMPLE_ID'].to_numpy()[0]\n",
    "        gdna = gdna[:gdna.rfind('_')]\n",
    "        if row['TUMOR'] != 0:\n",
    "            for elem in Path(PATH).rglob('{}*_vs_{}*/Control-FREEC/*pileup_ratio.txt'.format(sample_id, gdna)):\n",
    "                print(elem)\n",
    "                with open(elem, 'r') as filehandler:\n",
    "                    for line in filehandler.readlines()[1:]:\n",
    "                        tokens = line.split('\\t')\n",
    "                        chrm = str(tokens[0])\n",
    "                        start = int(tokens[1])\n",
    "                        ratio = float(tokens[2])\n",
    "                        # Control-FREEC defines a -1 ratio as not present\n",
    "                        if ratio != -1:\n",
    "                            cv = int(tokens[4])\n",
    "                            cnvs_chr['{}:{}'.format(chrm, start)][sample_id] = (cv,ratio)\n",
    "                \n",
    "# SECOND parse the dictionary to create a table with the CNVs\n",
    "ALL_SAMPLES = list(metasamples.loc[metasamples['TUMOR'] != 0, 'SAMPLE_ID'])\n",
    "ALL_PATIENTS = list(metasamples.loc[metasamples['TUMOR'] != 0, 'PATIENT_ID'])\n",
    "with open('../analysis/merged_cnvs.txt', 'w') as handler1:\n",
    "    with open('../analysis/merged_cnvs_ratios.txt', 'w') as handler2:\n",
    "        header_samples = '\\t'.join(['{}-{}'.format(x, y) for x, y in zip(ALL_SAMPLES, ALL_PATIENTS)])\n",
    "        handler1.write('CHR\\tPOS\\tGENE\\t{}\\n'.format(header_samples))\n",
    "        handler2.write('CHR\\tPOS\\tGENE\\t{}\\n'.format(header_samples))\n",
    "        for key, items in cnvs_chr.items():\n",
    "            samples_str_cv = '\\t'.join([str(items[s][0]) if s in items else 'Na' for s in ALL_SAMPLES])\n",
    "            samples_str_ratio = '\\t'.join([str(items[s][1]) if s in items else 'Na' for s in ALL_SAMPLES])\n",
    "            chrm, start = key.split(':')\n",
    "            gene_names = data.gene_names_at_locus(contig=chrm, position=int(start))\n",
    "            gene = ';'.join(gene_names)\n",
    "            to_write_cv = 'chr{}\\t{}\\t{}\\t{}\\n'.format(chrm, start, gene, samples_str_cv)\n",
    "            handler1.write(to_write_cv)\n",
    "            to_write_ratio = 'chr{}\\t{}\\t{}\\t{}\\n'.format(chrm, start, gene, samples_str_ratio)\n",
    "            handler2.write(to_write_ratio)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "cnvs = pd.read_csv('../analysis/merged_cnvs.txt', sep=\"\\t\", header=0, index_col=None, low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "data_dict = dict()\n",
    "for name, group in cnvs.iloc[:,2:].groupby('GENE', axis=0):\n",
    "    data = group.iloc[:,1:].transpose()\n",
    "    data[data == 'Na'] = np.nan\n",
    "    data = data.astype('float')\n",
    "    data_dict[name] = data.mean(axis=1)\n",
    "df = pd.concat(data_dict.values(), axis=1, keys=data_dict.keys())\n",
    "df.transpose().to_csv(\"../analysis/merged_cnvs_genes.txt\", sep=\"\\t\", index=True, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
